iceoryx_examples/icedelivery/iox_subscriber_untyped_modern.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/popo/modern_api/subscriber.hpp"
+#include "iceoryx_posh/popo/guard_condition.hpp"
+#include "iceoryx_posh/popo/wait_set.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "topic_data.hpp"
+
+#include <chrono>
+#include <csignal>
+#include <iostream>
+
+bool killswitch = false;
+iox::popo::GuardCondition shutdownGuard;
+
+static void sigHandler(int f_sig [[gnu::unused]])
+{
+    // caught SIGINT, now exit gracefully
+    killswitch = true;
+    shutdownGuard.trigger(); // unblock waitsets
+}
+
+void subscriberHandler(iox::popo::WaitSet& waitSet)
+{
+    // run until interrupted
+    while(!killswitch)
+    {
+        auto triggeredConditions = waitSet.wait();
+        for(auto& condition : triggeredConditions)
+        {
+            auto untypedSubscriber = dynamic_cast<iox::popo::UntypedSubscriber*>(condition);
+            if(untypedSubscriber)
+            {
+                untypedSubscriber->take().and_then([](iox::cxx::optional<iox::popo::Sample<const void>>& maybeSample){
+                    if(maybeSample.has_value())
+                    {
+                        auto position = reinterpret_cast<const Position*>(maybeSample->get());
```

Comment:
I am not sure and did not try it out, but it could define and_then to be a template<T> which will do the cast before passing it to the lambda. But then the lambda would need to take a T& (and, as previously suggested, only be called by and_then when there actually is data). 
Type inference would most likely not work based on the lambda type I think, so we would call something like  ```untypedSubscriber->take().and_then<T>(...)```. 

Having some API call such as ```Sample<T> typedSample = untypedSample.as<T>;``` which converts the untyped ```Sample<void>``` to a typed one may also be nice, if it can be done in a lightweight manner (i.e. no heavy copies). This would require some further thought to get it right (usable, efficient) though.

We then could do something like ```untypedSubscriber->take().as<T>().and_then(...)```. Maybe build it in take() as an alternative, i.e. ```untypedSubscriber->take<T>().and_then(...)```. The idea is to get a typed sample as early as possible to be able to use the typed API.

Note that this is not fully thought through, just some ideas. I would prefer to avoid having the user do too many reinterpret or dynamic casts (much boilerplate and error-prone).